### PR TITLE
[Feature] - Add RepeatWhen operator to Observable

### PR DESCRIPTION
**Pull request by sergdort**
This PR adds repeatWhen operator to observable sequence #396 

### Description of change

* Implement adopt almost all logic from the `RetryWhen` operator
* Add `RepeatWhenSequence`, `RepeatWhenSequenceSink`, `RepeatWhenSequenceSinkIter`, `RepeatTriggerSink`
* Add `repeatWhen` extension methods on `ObservableType`

### Questions

* Is it worth to include to the main repo? :) 
* I feel a little bit sorry for copy pasting almost everything from `RetryWhen`, so I'm thinking maybe there is a way to refactor so we can share logic between both of the operators ?

### CODE REVIEW

**Comment by sergdort** [comment]
I've ported tests from [RxJs](https://github.com/Reactive-Extensions/RxJS/blob/8fa95ac884181fb6cbff8ce7c1d669ffb190f5e4/tests/observable/repeatwhen.js)

**Comment by sergdort** [comment]
There are some tests missing regarding resource management. Gonna add them today

**Comment by kzaher** [comment]
@sergdort I've added them. No worries. It took a while to get this PR merged, so I've speeded it up a bit. I've made those comments FYI so you know what I've changed.

**Comment by sergdort** [comment]
@kzaher  I think you commenting into another PR 😜

**Comment by kzaher** [comment]
Oh, yeah. My bad :)

File: RxSwift/Observables/RepeatWhen.swift [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+extension ObservableType {
+    /**
+    Returns an Observable that emits the same values as the source Observable with the exception of an onCompleted. An onCompleted notification from the source will result in the emission of a void item to the Observable provided as an argument to the notificationHandler function. If that Observable calls onComplete or onError then repeatWhen will call onCompleted or onError on the child subscription. Otherwise, this Observable will resubscribe to the source observable.
+     
+     - parameter notificationHandler: receives an Observable of notifications with which a user can complete or error, aborting the repeat.
+     
+     - returns: An observable sequence producing the elements of the given sequence repeatedly
+    */
+    public func repeatWhen<TriggerObservable: ObservableType>(_ notificationHandler: @escaping (Observable<Void>) -> TriggerObservable)
```
Comment (sergdort) [inline RxSwift/Observables/RepeatWhen.swift]: Not sure if this closure should `throw` e.g `(Observable<Void>) throws ->  TriggerObservable `?

**Comment by sergdort** [comment]
I guess we can ignore `push` as I accidentally pushed to `upstream`

**Comment by kzaher** [comment]
Hi @sergdort ,

Let me first solve #1111 since I've started with it, then I'll take a more detailed look.

**Comment by sergdort** [comment]
@kzaher sure! Take your time :)

File: RxSwift/Observables/RepeatWhen.swift (hunk 1 above)
Comment (kzaher) [inline RxSwift/Observables/RepeatWhen.swift]: In `retryWhen` we don't have `throws` and nobody misses it :) I think it should be also ok here.

Comment (freak4pc) [inline RxSwift/Observables/RepeatWhen.swift]: Yeah I agree with here, it doesn't really fit within the other implementation or the usage of it :) I think it's good without.

**Comment by freak4pc** [comment]
Seems #1111 was solved and merged, do we want to go back to this ? :)

**Comment by gabornagyfarkas** [comment]
I'm also very interested in this feature, is it planned to be merged in the near future?

**Comment by kzaher** [comment]
4.0 is priority right now. After that we'll take a look at this PR since it doesn't block anything.

**Comment by fergalrooney** [comment]
I am also very interested in this feature. Is there still plans to merge into a release?

**Comment by BenGroot** [comment]
I'm converting Kotlin code to Swift and I'm missing the repeatWhen in RxSwift. Can you tell me anything about the progress on this?

**Comment by kaushikgopal** [comment]
@kzaher  : just wanted to follow up on this. Given  @sergdort has done the work to get us close, any plans on taking this to the finish line?

**Comment by MortyMerr** [comment]
Hey @sergdort, would you mind if I continue your work on this?

**Comment by freak4pc** [comment]
Moving the discussion over to #1990

**Pull request closed without merging**